### PR TITLE
feature: Add gemini backend

### PR DESCRIPTION
**Pull request by aislasq**
This is the implementation of the Gemini API. 

I've been using for general questions outside neovim and it works well, I think it would greatly benefit from the custom system prompts.

- [X] Implement trait for new backend.
- [X] Update the [BackendName](https://github.com/dustinblackman/oatmeal/blob/main/src/domain/models/backend.rs) enum with your new Backend name.
- [X] Update the [BackendManager](https://github.com/dustinblackman/oatmeal/blob/main/src/infrastructure/backends/mod.rs) to provide your new backend.
- [X] Write tests
- [X] Update Readme
- [X] Update config

### CODE REVIEW

**Review by dustinblackman — CHANGES_REQUESTED** [review]
Hey there! Thanks for the PR, this is really exciting. I haven't had the chance to play with Gemini yet, and it'll be sick for my first experience to be in Oatmeal. 

Just got a few comments and questions, mostly around getting tests fixed up.

File: src/infrastructure/backends/gemini.rs [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+            bail!("Gemini URL is not defined");
+        }
+        if self.token.is_empty() {
+            bail!("Gemini token is not defined");
+        }
+
+        let url = format!(
+            "{url}/v1beta/{model}?key={key}",
+            url = self.url,
+            model = Config::get(ConfigKey::Model),
```
[the remark this answers is not in the document]
Comment (aislasq) [reply src/infrastructure/backends/gemini.rs:116]: Is there a possibility to get more info on the test? I am doing the Config::set in the test exactly as you mentioned:
https://github.com/aislasq/oatmeal/blob/5f4fd4f89b0634060e456ef9b3e4144167ddfeee/src/infrastructure/backends/gemini_test.rs#L54

File: src/infrastructure/backends/gemini.rs [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+
+#[derive(Default, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct Model {
+    name: String,
+    // version: String,
+    // display_name: String,
+    // description: String,
+    // input_token_limit: u32,
+    // output_token_limit: u32,
```
[the remark this answers is not in the document]
Comment (aislasq) [reply src/infrastructure/backends/gemini.rs:39]: Yes, but we can remove them and add them later when needed.

File: src/infrastructure/backends/gemini_test.rs [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    mock.assert();
+
+    assert_eq!(res, vec!["first".to_string(), "second".to_string()]);
+
+    return Ok(());
+}
+
+#[tokio::test]
+async fn it_gets_completions() -> Result<()> {
+    let body = [
```
[the remark this answers is not in the document]
Comment (aislasq) [reply src/infrastructure/backends/gemini_test.rs:131]: Believe it or not, Google's stream implementation is sending a 'prettified' json object line by line. If you take a closer look I am not serializing the object, I am appending it by line to get something like this:
```json
{
  "contents": [
    {
      "parts" : [{
        "text": "Hello"
      }]
    },
    {
      "parts": [{
        "text": "World"
      }]
    }
  ]
}
```

File: config.example.toml [hunk 4]
```diff
@@ -22,6 +22,9 @@ ollama-url = "http://localhost:11434"
 # OpenAI API URL when using the OpenAI backend. Can be swapped to a compatible proxy.
 open-ai-url = "https://api.openai.com"
 
+# Gemini API token when using the OpenAI backend.
```
Comment (aislasq) [inline config.example.toml:25]: Must fix before merging

**Comment by aislasq** [comment]
The lint is failing because my formatter is changing it on save, I am using rustfmt 1.6.0-stable (79e9716c 2023-11-13). Should I use a different one?

File: src/infrastructure/backends/gemini.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (aislasq) [reply src/infrastructure/backends/gemini.rs:116]: I know what's going on, I'm an idiot. Of course it's going to fail, it expects an env variable. Give me a minute to check if I can generate a new one so you can add it to the secrets in the repo.

File: src/infrastructure/backends/gemini_test.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (aislasq) [reply src/infrastructure/backends/gemini_test.rs]: Changed the array to a raw string. 
This is the official documentation:
https://ai.google.dev/tutorials/rest_quickstart#stream_generate_content

File: src/infrastructure/backends/gemini.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (dustinblackman) [reply src/infrastructure/backends/gemini.rs:108]: The github actions test suite is using `nextest`, this runs tests in parallel process' where memory will be different for each test. You have `Config:set` being used in one test, but it's gotta be use in all your [healthcheck tests](https://github.com/aislasq/oatmeal/blob/5f4fd4f89b0634060e456ef9b3e4144167ddfeee/src/infrastructure/backends/gemini_test.rs#L74C10-L74C32).

**Comment by dustinblackman** [comment]
> The lint is failing because my formatter is changing it on save, I am using rustfmt 1.6.0-stable (79e9716c 2023-11-13). Should I use a different one?

The project is using a nightly version of rustfmt to gain access to other features I wanted. You can run `cargo cmd lint-fix` to fix everything before committing. If the command fails, you may need to take a look at the [setup guide](https://github.com/dustinblackman/oatmeal?tab=readme-ov-file#setup) to get all the dependencies setup, otherwise I should have a devcontainer available tomorrow if that's a workflow you enjoy using. :)

File: src/infrastructure/backends/gemini.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (aislasq) [reply src/infrastructure/backends/gemini.rs:108]: Done, let me check the linter so you can re-run the tests. Thanks!

**Review by dustinblackman — APPROVED** [review]
Really appreciate this, code looks great and I'm excited to play with it more. Great job!

**Pull request merged**